### ISSUE TITLE
How to create a new image based on this

### ISSUE THREAD

**Issue by seriema** [lukechilds/dockerpi#3]
Hi!

Running this image works great! Unfortunately I creating my own image based on it was less straightforward than I thought. I'm new to Docker but a Dockerfile like this:

```
FROM  lukechilds/dockerpi

RUN ls -la
```

outputs a folder structure that doesn't look like a Raspberry Pi.
```
drwxr-xr-x    1 root     root          4096 Jan 12 19:21 .
drwxr-xr-x    1 root     root          4096 Jan 12 19:21 ..
-rwxr-xr-x    1 root     root             0 Jan 12 19:21 .dockerenv
drwxr-xr-x    2 root     root         12288 Dec  2 20:12 bin
drwxr-xr-x    5 root     root           340 Jan 12 19:21 dev
-rwxr-xr-x    1 root     root           672 Dec 16 22:54 entrypoint.sh
drwxr-xr-x    1 root     root          4096 Jan 12 19:21 etc
-rw-------    1 root     root     454279954 Sep 26 00:26 filesystem.zip
drwxr-xr-x    2 nobody   nogroup       4096 Dec  2 20:12 home
dr-xr-xr-x  189 root     root             0 Jan 12 19:21 proc
drwx------    1 root     root          4096 Dec 16 22:59 root
drwxr-xr-x    2 root     root          4096 Jan 12 19:20 sdcard
dr-xr-xr-x   13 root     root             0 Jan 12 19:21 sys
drwxrwxrwt    1 root     root          4096 Dec 16 22:59 tmp
drwxr-xr-x    1 root     root          4096 Dec 16 22:59 usr
drwxr-xr-x    4 root     root          4096 Dec  2 20:12 var
```

This image would really be useful if the Readme could give an example Dockerfile showing how to create images based on it.

**Comment by lukechilds** [comment]
I think you're misunderstanding what this image is doing.

It's not a Docker image of a Raspberry Pi like system.

It's a Docker image of a minimal system that runs QEMU. QEMU then virtualises a Raspberry Pi like system and boots `/filesystem.zip`.

What's your use case for using this as a base image?

If you want to boot a custom system than you can pass in your own filesystem to be booted:

```
docker run -it -v /2019-09-26-raspbian-buster-lite.img:/sdcard/filesystem.img lukechilds/dockerpi:vm
```

**Comment by seriema** [comment]
@lukechilds Hi, sorry for the late reply.

I'd like to create a [RetroPie](https://retropie.org.uk/) image for my little project [Retro-Cloud](https://github.com/seriema/retro-cloud). It would enable me to develop locally on my laptop (without connection to my rpi at home), and set up a build pipeline to verify my commits (and PR's).

My project downloads and installs some needed tools (like PowerShell and it's Azure module) to create some online storage for games etc. Which means I often need to re-run the scripts to see if they're working as expected. Without an image this is a pain. I _could_ run docker on the rpi, which is what many seem to be doing. But I'd like to use my laptop and that's what I'd imagine I'd use your project for.

That means step 1 in my Dockerfile would be [the RetroPie installation steps](https://retropie.org.uk/docs/Manual-Installation/). I'd hope for some simple example in the Readme on how to build on top of your image. Does this make sense?

**Comment by aaronpropst** [comment]
Yeah, what your dockerfile would be modifying would be the emulation environment.  Akin to modifying the pi, not the image.   

You'd want to rig your build process to fire up this image and run your build script against an off the shelf image modifying and customizing it with your changes.

**Comment by seriema** [comment]
Just installing RetroPie when running your image takes an hour. That’s why I’d like the image to start from it already installed. Your emulation setup is precisely what I want and I imagine it being really powerful for others to create images with what they want preinatalled on an rpi and running it emulated. But perhaps I don’t understand your intentions with this project?

**Comment by aaronpropst** [comment]
The limiting factor there may be the resources you've given docker, or possibly the resources Luke has given qemu.   
for instance, ram seems to be specified on this line:
https://github.com/lukechilds/dockerpi/blob/master/entrypoint.sh#L20

Hacking around with those values may let it install faster,  and once you install, it will then be installed on that image and not need to be installed again.  (I haven't played too deeply with this mind you, I only did a test that verified that files I changed stayed changed in the image across docker runs)

I haven't figured out how to automatically script changes to a vanilla image yet though, as this entrypoint drops you to a login prompt as opposed to a running raspbian system.   My best early kludge idea would be to put a 'ssh' file in the boot directory of the image, so ssh is available, then ssh to the running raspbian and run install scripts.

Automating raspbian setup from the /boot directory is something I and others have long lobbied for.. : (
https://www.raspberrypi.org/forums/viewtopic.php?t=131443

**Comment by aaronpropst** [comment]
As for intentions.. I can't speak for the author, but for me at least, it removes the necessity of installing and setting up Qemu..  not just for me, but for those I'd communicate with.. I could for instance give someone a docker compose file referencing this image and a pi image, and they could be playing around on a virtualized pi image I've modified after executing one command.

I'm churning on other purposes as well (such as a build process above)  I've always wanted a declarative way to define raspbian setups where every change you've made from the distributed image can be committed in one little repo.  (similar to how docker allows you to bake all your system changes into a neat dockerfile)

I suspect you're after the same thing, but this isn't quite it.  Not without a bit more innovation.

**Comment by lukechilds** [comment]
Yeah, @aaronpropst pretty much hit the nail on the head.

Also with regards to performance, I'm currently using the versatilepb machine which is quite limited. It only allows up to 256MB of RAM and I'm virtualising an arm1176 CPU which is a low powered single core 32 bit CPU like in the original Raspberry Pi.

QEMU does have a raspi3 machine which would allow us to virtualise a modern 64 bit quad core CPU and up to 1GB of RAM. However I haven't had any luck booting Raspbian with this machine. I asked for help here: https://www.raspberrypi.org/forums/viewtopic.php?p=1581855#p1581855 but haven't had a response yet.

If you need better performance than getting the raspi3 or another machine working and booting Raspbian is what you should aim for. If you can get that working and post the QEMU settings and any custom kernels required I can update this project to do the same.

**Comment by adamwolf** [comment]
I want to say thanks to everyone involved, certainly @lukechilds :)

I'd love to run some ansible playbooks against the QEMU target as part of a CI/CD thing.

Any thoughts on getting the IP out?

**Comment by herzogf** [comment]
Dockerpi already exposes the qemu/raspbian ssh port on port 5022 in the docker container. You could access it with the container's ip or publish the port on the host (docker run -p 127.0.0.1:5022:5022 ...) so that your CICD scripts can access it with localhost:5022 instead of the dynamic container ip.

Lastly as mentioned above you have to activate SSH in raspbian so that ansible can connect to it. I did this by mounting the raspbian image upfront and adding a ssh file to the boot partition, see https://github.com/herzogf/homek8s/blob/a882aceb24b9e522db374d344dee1089b4f32e7d/test/native_install_prereqs.sh#L27-L33
Then you should be able to use this image in dockerpi like documented in the dockerpi docs:
```bash
docker run -d -p 127.0.0.1:5022:5022 -v /myCustomImageWithSshFile.img:/sdcard/filesystem.img lukechilds/dockerpi
```

**Comment by dbaid** [comment]
Hello lukechilds ,

   Do you consider bridge network ??

Regards,
Steven
